aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -57,6 +57,54 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
   "The input tensor must be a matrix with size ", self.sizes()[0], "x", mat2.sizes()[1], ", but got a ", result.dim(),
   "-D tensor with size ", result.sizes()[0], "x", result.sizes()[1]);
 }
+
+template <typename Meta>
+void common_checks_baddbmm_bmm(Meta& meta, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha, bool is_bmm, const Tensor& self_or_result) {
+  TORCH_CHECK(batch1.dim() == 3, "batch1 must be a 3D tensor");
+  TORCH_CHECK(batch2.dim() == 3, "batch2 must be a 3D tensor");
+
+  const auto batch1_sizes = batch1.sizes();
+  const auto batch2_sizes = batch2.sizes();
+
+  int64_t bs = batch1_sizes[0];
+  int64_t contraction_size = batch1_sizes[2];
+  int64_t res_rows = batch1_sizes[1];
+  int64_t res_cols = batch2_sizes[2];
+
+  TORCH_CHECK(batch2_sizes[0] == bs && batch2_sizes[1] == contraction_size);
+
+  if (!is_bmm) {
+    TORCH_CHECK(self_or_result.dim() == 3, "self must be a 3D tensor");
+    const auto self_sizes = self_or_result.sizes();
+    TORCH_CHECK(self_sizes[0] == bs && self_sizes[1] == res_rows && self_sizes[2] == res_cols);
+  }
+
+  namedinference::propagate_names_if_nonempty(
+    self_or_result,
+    namedinference::compute_bmm_outnames(const_cast<Tensor&>(self_or_result), batch1, batch2)
+  );
+}
+
+TORCH_META_FUNC(bmm)(const Tensor& self, const Tensor& mat2) {
+    set_output({self.sizes()[0], self.sizes()[1], mat2.sizes()[2]}, self.options());
+    auto& result = maybe_get_output(0);
+    common_checks_baddbmm_bmm(*this, self, mat2, Scalar(0.0), Scalar(1.0), true, result);
+}
+
+TORCH_META_FUNC(baddbmm)(const Tensor& self, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha) {
+  auto& result = maybe_get_output(0);
+  if (!result.defined()) {
+    set_output({0}, batch1.options());
+  }
+  if (!result.is_same(self)) {
+    set_output({batch1.size(0), batch1.size(1), batch2.size(2)}, self.options());
+    if (beta.to<c10::complex<double>>() != 0.0) {
+      result.copy_(self);
+    }
```

Comment:
I think this belongs to the `IMPL` function (it's neither a check nor output metadata setting).

Reply:
Then, you would have to replicate it to the CUDA `IMPL` function, too.

Reply:
The reason it's here is that the size of `self` is modified using `expand_size` (not in-place though, we create another tensor and then use it from there for the checks).